components/eventing-controller/controllers/subscription/nats/reconciler.go

```diff
@@ -173,139 +172,201 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
+				events.Warn(r.recorder, subscription, events.ReasonUpdateFailed, "Update Subscription failed %s", subscription.Name)
 				log.Errorw("remove finalizer from subscription failed", "error", err)
+				if k8serrors.IsConflict(err) {
+					return ctrl.Result{Requeue: true}, nil
+				}
 				return ctrl.Result{}, err
 			}
 			log.Debug("remove finalizer from subscription succeeded")
-			return ctrl.Result{}, nil
 		}
-	}
-
-	// Check for valid sink
-	if err := r.sinkValidator(ctx, r, desiredSubscription); err != nil {
-		log.Errorw("sink URL validation failed", "error", err)
-		if err := r.syncSubscriptionStatus(ctx, actualSubscription, false, err.Error()); err != nil {
-			return ctrl.Result{}, err
-		}
-		// No point in reconciling as the sink is invalid
+		// Stop reconciliation as the object is being deleted
 		return ctrl.Result{}, nil
 	}
 
 	// The object is not being deleted, so if it does not have our finalizer,
 	// then lets add the finalizer and update the object. This is equivalent
 	// registering our finalizer.
-	if !utils.ContainsString(desiredSubscription.ObjectMeta.Finalizers, Finalizer) {
-		desiredSubscription.ObjectMeta.Finalizers = append(desiredSubscription.ObjectMeta.Finalizers, Finalizer)
-		if err := r.Update(context.Background(), desiredSubscription); err != nil {
+	if !utils.ContainsString(subscription.ObjectMeta.Finalizers, Finalizer) {
+		subscription.ObjectMeta.Finalizers = append(subscription.ObjectMeta.Finalizers, Finalizer)
+		if err := r.Update(context.Background(), subscription); err != nil {
 			log.Errorw("add finalizer to subscription failed", "error", err)
+			if k8serrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
 			return ctrl.Result{}, err
 		}
 		log.Debug("add finalizer to subscription succeeded")
-		result.Requeue = true
+		return ctrl.Result{Requeue: true}, nil
 	}
 
-	if result.Requeue {
-		return result, nil
+	// Check for valid sink
+	if err := r.sinkValidator(ctx, r, subscription); err != nil {
+		log.Errorw("sink URL validation failed", "error", err)
+		if err := r.syncSubscriptionStatus(ctx, subscription, false, false, err.Error()); err != nil {
+			if k8serrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, err
+		}
+		// No point in reconciling as the sink is invalid
+		return ctrl.Result{}, nil
 	}
 
-	_, err = r.Backend.SyncSubscription(desiredSubscription, r.eventTypeCleaner)
+	// Synchronize Kyma subscription to NATS backend
+	subscriptionStatusChanged, err := r.Backend.SyncSubscription(subscription, r.eventTypeCleaner)
 	if err != nil {
 		log.Errorw("sync subscription failed", "error", err)
-		if err := r.syncSubscriptionStatus(ctx, desiredSubscription, false, err.Error()); err != nil {
+		if err := r.syncSubscriptionStatus(ctx, subscription, false, false, err.Error()); err != nil {
+			if k8serrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
 			return ctrl.Result{}, err
```

Comment:
While I really like the idea of using `ctrl.Result{Requeue: true}` whenever there is a k8s conflict error, I see that it bloats the code up a lot. I don't see the case that the caller of `syncSubscriptionStatus` wants to perform custom error handling. So maybe we should move this to the callee instead and save some lines of code to reduce the size of the methods. What is your take on this ?

Reply:
handle it in a new helper method